### ISSUE TITLE
Suggest changing your escape code mechanism

### ISSUE THREAD

**Issue by CGamesPlay** [sedwards2009/extraterm#20]
Hi, just saw you starred my repo so I wanted to check this one out as well. I think you should change the escape code you're using for HTML output. I'll put this into a design document over at my repo soon but in the mean time:
- You should use a custom DSR to decide if your terminal supports HTML or not. We could standardize on one like DSR 1866, then you can use a [shell script like this](https://github.com/gnachman/iTerm2/blob/master/tests/isiterm2.sh) to decide to enable your integrations or not.
- You should use an OSC to read a complete HTML blob at a time, rather than having the output stream wait until a NUL byte. An OSC is strictly a sequence of printable characters, meaning that if your HTML program crashes, the next escape sequence printed (e.g. color code) will cause the OSC to be invalid, resulting in HTMl getting dumped to the terminal as text but critically not ruining the terminal for other programs who aren't aware of HTML.

Since you support outputting HTML and so do I, I definitely think it makes sense to standardize the escape codes we use. I picked 1866 because it wasn't in use and is a reference to RFC 1866.

**Comment by sedwards2009** [comment]
Right now the code to dump HTML into the terminal is commented out in the Extraterm code. The reason being that I need to have a good think about how it should work and what it actually should do to be really useful.

Some questions that need better answers:

* Terminals are a keyboard oriented interface. How does that mesh with HTML which is often more mouse oriented? How can we make any HTML shown in the terminal keyboard friendly?
* Extraterm's cursor/selection mode which lets you copy stuff and move around the screen and scrollback via the keyboard is actually extremely useful. How would this mode interact with a HTML display mode?

Another very big consideration is figuring out how to make this kind of functionality secure, safe and robust. For example, if you use `cat` to show the contents of a file then NONE of the escape codes should be available for use. Always accepting escape codes would make it possible for people to embed HTML and Javascript code in a file and use it to attack the user's via the terminal itself when someone views it using `cat` or `less`. This is why Extraterm's escape codes require a cookie value to be present before they are accepted. This cookie value is made available to terminal applications as an environmental variable.

I would be happy to see some kind of standardisation of this kind of functionality in the future, but there is an awful lot of work and experimentation which has to be done first.

**Comment by CGamesPlay** [comment]
>Terminals are a keyboard oriented interface. How does that mesh with HTML which is often more mouse oriented? How can we make any HTML shown in the terminal keyboard friendly?

Arrows to change focus and space to activate buttons doesn't seem like the worst, but you're right. I think the mouse interfaces are a nice to have for more complex apps, but the bulk of the value is in richer visuals for components (`less` replaced with a text overflow div, but all the keyboard controls for less are still intact).

>Extraterm's cursor/selection mode which lets you copy stuff and move around the screen and scrollback via the keyboard is actually extremely useful. How would this mode interact with a HTML display mode?

This actually looks really cool. You could maybe use `contenteditable` to make the HTML region editable, although I'm not sure it's as useful as text-mode commands being editable, because I imagine programs that output HTML are probably unlikely to produce output suitable for being input to other programs.

> Another very big consideration is figuring out how to make this kind of functionality secure, safe and robust.

Ah, cookie makes sense then. There are 2 key principles that need to be adhered to:

1. Program output must not cause the terminal to send input based on previous output the terminal has received (this applies to all terminal programs; think about `ssh` being able to read data out of your scrollback).

2. Program output must not cause the terminal to transmit data over the network (this is unique to HTML-enabled terminals).

The way to safely accomplish both of these is to parse any received HTML and transform it into "safe" HTML, using a whitelist of tags and attributes. For me, this is going to be basic HTML tags and a limited set of attributes like `class` or image's `src` but restricted to data URIs (but not `style`, because that could have URLs). For richer components (I'm thinking like `progress-bar`), those are bundled with the terminal program (or installed by the user as extensions). This last set allows you to have rich components without needing to allow programs to use JavaScript.

This works out rather nicely: if you overwrite an HTML blob (think clearing the screen and rewriting it), React will detect this as a prop change rather than a delete and recreate, meaning that for no extra work my `progress-bar` component can smoothly animate between different states, rather than flickering.

I have not implemented any of the security features I talk about above, but the code to do that is going to look at lot like [HTMLParser.js](https://github.com/CGamesPlay/hterminal/blob/master/client/HTMLParser.js). If you're interested, we could split off that file into a module suitable for parsing "Terminal-safe HTML" and both share it?

**Comment by sedwards2009** [comment]
Extraterm is more or less powered by CodeMirror. That is where the cursor, selection and editing capabilities come from. I was thinking of doing something similar for HTML and rich text formats. Integrating ProseMirror or some other WYSIWYG editing into Extraterm and using it to display the content and handle cursors and selections etc.

As for displaying 'application' HTML safely, I lean towards sandboxing instead of trying to clean up HTML. Web browsers work the same way when displaying or running a web page.
